### ISSUE TITLE
Bulk update

### ISSUE THREAD

**Issue by naormatania** [ent/ent#1815]
This feature proposal is about updating many entities in a single query similar to how `Create Many` and `Upsert Many` work.

I'd love to hear people thoughts about if this proposal sounds helpful to others, if there could be challenges I don't anticipate and what's the best way to support it (especially since it might be better to only support it in postgres, details below) - external templates ? contribution to the project ? feature flag ?

- [x] I have searched the [issues](https://github.com/ent/ent/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

In SQL there are two ways to support "bulk update".
The first way that is supported in both MySQL and Postgres is to use "cases" with values for each ID and put the current value if no new value is needed:
```sql
UPDATE grades SET grade = (CASE id WHEN 1 THEN 30
WHEN 2 THEN 80
ELSE grade
END),
course_id = (CASE id WHEN 1 THEN 'Math'
ELSE course_id
END),
WHERE id IN (1, 2)
```
However this way creates a very large query that would introduce performance penalties in different levels.

The second way is only supported in Postgres and it is to use the `UPDATE ... FROM ...` syntax that allows updating rows from "temporary table" directly.
The syntax for the query could be like:
```sql
update grades 
set course_id = data_table.course_id,
student_id = data_table.student_id,
grade = data_table.grade
from 
  (WITH vals (id, course_id, student_id, grade) as (VALUES (1, 'Math', 1000, 95), (2, 'Math', 1001, 100)) SELECT * from vals) as data_table
where grades.id = data_table.id;
```
Another way to write the query is to use postgres array (which might be better since `VALUES` should have some limit from what I understand):
```sql
  update grades g
  set course_id = data_table.course_id,
  student_id = data_table.student_id,
  grade = data_table.grade
  from 
    (select unnest(array[1,2]) as id, unnest(array[g.course_id, 'Math2']) as course_id, unnest(array[1000, 1001]) as student_id, unnest(array[95, g.grade]) as grade) as data_table
  where grades.id = data_table.id;
```
These queries are very fast and scales way better than the first way I presented. The only issue with them is when I want to update different columns in different rows (like I do in the first way I presented). Possible solution is shown in https://stackoverflow.com/questions/68670268/is-it-possible-to-bulk-update-specific-values-in-postgresql-efficiently, however it requires to rely on "impossible value" for each column which is not ideal.
I address it in the API proposal I show next by stating the updated fields in the `BulkUpdate` API:

```golang
bulk := make([]*ent.PetUpdateOne, 2)
bulk[0] = client.Pet.UpdateOne(pet1).SetName("pedro").SetAge(10)
bulk[1] = client.Pet.UpdateOne(per2).SetName("layla").SetAge(13)
pets, err := client.Pet.UpdateBulk(bulk...).
              SetUpdatedFields(pet.FieldName, pet.FieldAge).
              Save(ctx)
```
The new API `UpdateBulk` receives a list of `UpdateOne` builders and also receives a list of expected update fields that should exist with each of the builders.
The order of the execution when `UpdateBulk` runs:
1. Hooks of `UpdateOne` are called for each of the builder (which might set more fields like `update_time` for example`)
2. A "low level hook" of `UpdateBulk` is called that checks that each builder has set exactly the list of update fields (not more than those fields and not less), if not it fails the API call if not
3. A query is constructed (according to way (1) or (2) I presented) and sent

## Motivation 🔦

When building pipeline system you usually receive batch of results that you wish to either create or update in database efficiently (i.e in less queries as possible).
One way to support this use case is using the bulk upsert feature (https://entgo.io/docs/crud#upsert-many). However this feature is limited in several ways:
- Sometimes you wish to update an existing rows based on the new value and their current values like doing max/sum between them that can potentially be added to the upsert query (that complicates the query though) or advanced calculations that simply cannot be done in sql
- Hooks cannot be run for update operation
- If there is a logic that relies on whether create or update "events", it becomes difficult to understand the "event" that has happened.

Therefore, I'd like to have "bulk update" so the logic of pipeline could be the following - First, select rows to find out which rows needs to be created and which to be updated (query 1), bulk create all rows that needs to be created (query 2) and bulk update all rows that needs to be updated (query 3).
While this operation takes more queries than upsert, it is still quite efficient when handling large batch of results and it mitigates all the issues with upsert.

**Comment by liszt831020** [comment]
👀

**Comment by alexfreska** [comment]
👀

**Comment by pzeinlinger** [comment]
:eyes:

**Comment by mihalydanis** [comment]
Any update on this?

**Comment by thoriqadillah** [comment]
I think it is available on the docs already https://entgo.io/docs/crud/#update-many

**Comment by thoriqadillah** [comment]
I think it is available on the docs already https://entgo.io/docs/crud/#update-many
I believe the update many operation is just update operation with `where` condition, thus if the (multiple) data that presents on the table has met the condition, it will update all the said data

**Comment by mihalydanis** [comment]
> I think it is available on the docs already https://entgo.io/docs/crud/#update-many I believe the update many operation is just update operation with `where` condition, thus if the (multiple) data that presents on the table has met the condition, it will update all the said data

The `UpdateMany` you mentioned is not a suitable resolution for this missing functionality. It should be integrated using a similar API as `CreateBulk` (as mentioned in the initial proposal), allowing us to update any values of individual rows across any table in a single command.

**Comment by thoriqadillah** [comment]
> allowing us to update any values of individual rows across any table in a single command.

can you please give some example?
right now in my head, if i want to query update with mulitple rows, i would just use `update where in` . at least that is my workaround. (quick lookup on google) I found that other ORM such as Eloquent from Laravel is doing the same

**Comment by dmateusp** [comment]
@thoriqadillah I can give an example as I'm hitting this issue right now

I'm writing a script to change IDs to other IDs based on a mapping I have:
```
foo -> toto
bar -> tata
baz -> titi
etc ...
```

I can't do a bulk update using `Update(...).Where(...).` because I want to set a different value for each one of the rows I'm updating. 

I would like to use an `UpdateBulk`, to reduce the amount of network calls to the database (I have a migration script to run)

**Comment by Betula-L** [comment]
> @thoriqadillah I can give an example as I'm hitting this issue right now
> 
> I'm writing a script to change IDs to other IDs based on a mapping I have:
> 
> ```
> foo -> toto
> bar -> tata
> baz -> titi
> etc ...
> ```
> 
> I can't do a bulk update using `Update(...).Where(...).` because I want to set a different value for each one of the rows I'm updating.
> 
> I would like to use an `UpdateBulk`, to reduce the amount of network calls to the database (I have a migration script to run)

Yes, i wanna update multi line in one transaction, one network call, rather than multi, as described below,
`
update x set data='a' where id =1;
update x set data='b' where id =2;
update x set data='c' where id =3;
`
It's seems that ent do not support now.

**Comment by phamvinhdat** [comment]
I see this lib supports bulk-update: https://bun.uptrace.dev/guide/query-update.html#bulk-update

**Comment by migueldv90** [comment]
My main ORM is ENT, but I had to use Bun ORM for a few features that ENT is missing.

**Comment by XiaoK29** [comment]
![Image](https://github.com/user-attachments/assets/ef87fd4d-bdfb-431e-972c-55425d8400b4)
Do you currently support batch updates? I am currently facing a similar problem, where conflicts arise when using batch creation. If an incorrect ID is passed in to the frontend, it will be created, and then I will have to query the database again to verify its validity
